src/storage/src/render/persist_sink.rs

```diff
@@ -93,54 +95,56 @@ pub fn render<G>(
                     // (100 was chosen arbitrarily), and avoid having to make a batch
                     // per-timestamp.
                     for (row, ts, diff) in buffer.drain(..) {
-                        stashed_batches
-                            .entry(ts)
-                            .or_insert_with(|| {
-                                // TODO: the lower has to be the min because we don't know
-                                // what the minimum ts of data we will see is. In the future,
-                                // this lower should be declared in `finish` instead.
-                                write.builder(100, Antichain::from_elem(Timestamp::minimum()))
-                            })
-                            .add(&SourceData(row), &(), &ts, &diff)
-                            .await
-                            .expect("invalid usage");
+                        if write.upper().less_equal(&ts) {
```

Comment:
This is just an optimization, correct? If we naively wrote all data I'd expect it to still work. Or did you observe problems without this?

Reply:
It's not, we have to filter out data that is not beyond the write upper because we're doing a `compare_and_append` below

Reply:
Ah, I think I made a mistake there. This should be an `append()`, otherwise it will not work when multiple processes are trying to write for the same source. :see_no_evil:

Reply:
but there should be only one ingestion process that handles a particular source, no? The code below panics if the upper expectation of `compare_and_append` panics to leave whoever raced ahead to continue ingesting

Reply:
Just realized that I had an `append()` here (and `persist_sink.rs` in compute still has) and this was changed. The initial idea for the design was that we wanted to allow multiple writers, so that we can write with multiple processes at the same time for HA. With `compare_and_append` it's more brittle, because it can also compete with a past incarnation of itself that hasn't properly shut down yet.